spectec/src/backend-prose/render.ml

```diff
@@ -139,280 +331,206 @@ let rec render_iter env = function
   | Al.Ast.Opt -> "^?"
   | Al.Ast.List -> "^\\ast"
   | Al.Ast.List1 -> "^{+}"
-  | Al.Ast.ListN (expr, None) -> "^{" ^ render_expr env true expr ^ "}"
+  | Al.Ast.ListN (expr, None) -> "^{" ^ render_expr env expr ^ "}"
   | Al.Ast.ListN (expr, Some name) ->
-      "^{(" ^ render_name name ^ "<" ^ render_expr env true expr ^ ")}"
+      "^{(" ^ render_name name ^ "<" ^ render_expr env expr ^ ")}"
 
 and render_iters env iters = List.map (render_iter env) iters |> List.fold_left (^) ""
 
+
 (* Expressions and Paths *)
 
-and render_expr env in_math expr =
+(* fallback when expr cannot be embedded in a math block *)
```

Comment:
Can you add a comment how this can occur? Can it legally happen, or is it indicative of some internal error?

Reply:
I missed out on explaining this, sorry for the confusion. In short, it is legal for this to happen, and not an error.

There are certain AL expressions that must be rendered using plaintext, for example `TopLabelE`, which should be rendered as “a label is on the top of the stack.” Also, some may contain subexprs, like `ContE e`, rendered as “the continuation of … some rendered string of e ….” 

Considering this, `render_expr` works as follows:
- First try to translate AL expr to EL exp with `al_to_el_exp`.
- If the translation succeeds, (indicated by Some), render it using the Latex backend.
    - This translation succeeds **if and only if** the AL expr and all its subexprs are able to be rendered in `:math:` blocks.
- Otherwise (indicated by None), it must be the case that the expr itself or one of its subexpr cannot be rendered in `:math:`. So we  fall back to the original prose renderer in `render_expr’`.

Reply:
That makes sense, but even if the whole expressions cannot, shouldn't the subexpressions that _can_ still be rendered with the Latex backend? In particular, I would expect that none of the cases for Var, Num, Bool, names, operators (other than Not), accessors, paths, records, iterators, etc., would ever be rendered here directly.

I would assume there must be an invariant somehow that none of the EL-handled cases can contain a subexpression that cannot be rendered as EL.(*) Or is there a counter-example? If so, shouldn't it be necessary to somehow transform it away, e.g., by introducing an auxiliary variable? For example, consider a record with a nested `ContE`, wouldn't that produce very jumbled output otherwise?

(*) FWIW, that would also imply that the option monad shouldn't be necessary, since it can only legally fail at the outermost level already.

Reply:
**shouldn't the subexpressions that can still be rendered with the Latex backend?**

Yes, `render_expr'` calls `render_expr` for rendering its subexpressions. For instance,

```ocaml
and render_expr' env expr =
  match expr.it with
  ...
  | Al.Ast.UnE (op, e) ->
        let sop = render_al_unop op in
        let se = render_expr env e in
        sprintf "%s %s" sop se
```

**I would assume there must be an invariant somehow that none of the EL-handled cases can contain a subexpression that cannot be rendered as EL.**

There exists a counterexample for AL `IterE`. `IsDefinedE` can be its subexpression, for instance, `IterE (DefinedE (CallE (default, [ CallE (unpacktype, [ VarE (zt) ]) ])), [zt], *)` which should be rendered as `for all zt*, $default($unpacktype(zt)) is defined`.

FYI, this prose expression is from the execution semantics of `STRUCT.NEW_DEFAULT`, in a step saying: `5. Assert: Due to validation, for all zt*, $default($unpacktype(zt)) is defined.` This is generated from the IL sideconditions.

Other than `IterE`, the prose being generated now can be distinguished of EL-handled or not at the top level AST. Nevertheless, I am unsure that such invariant (modulo `IterE`) can persist to hold as the spec evolves, so I left `render_expr'` as close to the original logic as a foolproof.

Reply:
Right, but taking the UnE case as an example, that only makes sense for operators that are themselves in text form. In some of the other cases, the current algorithm would produce math operators outside math, some of that may not even be valid Latex (for example, `\cdot` is not allowed outside math mode, likewise things like `^?`).

In other places, parts of a formula are rendered in a split fashion. For example,
```ocaml
 | Al.Ast.AccE (e, p) ->
       let se = render_expr env e in
       let sp = render_path env p in
       sprintf "%s%s" se sp
```
This will not type-set correctly, there needs to be a single math wrapper around both, not two separate ones. However, I don't understand how such cases can even legally arise: if one of the two parts cannot be converted to EL, then that supposedly means that it will be rendered as prose, but then concatenating them without space as if they were formulas doesn't seem right.

That's why there needs to be some invariant. There are three classes of expressions:

1. EL-like expressions only containing EL expressions ⇒ rendered by Latex backend
2. AL-only expressions, possibly containing EL expressions ⇒ rendered by prose backend except for inner EL
3. Pseudo EL-like expressions containing AL expressions ⇒ must be rendered by prose backend, as prose

In particular, all expressions in the 3rd category must be rendered as real prose, since they can contain nested prose. It never makes sense to render them as a formula. For example, if the prose renderer were to encounter an expression like `ListE [ContE e1, ContE e2]`, then it would currently generate something like "[the continuation of e1, the continuation of e2]", which is not what we want.

There are only two meaningful possibilities for handling such an example:

* either it's considered illegal, i.e., no AL-specific expressions can ever occur inside a list, but then `render_expr'` should not handle lists at all

* or it can legally occur, but then `render_expr'` needs to render the list as proper prose, e.g., something like "the sequence consisting of ..., ..., and ..."

Under no circumstances should `render_expr'` render any node as a formula-like object. That doesn't compose with nested prose.

Effectively, there are two layers of languages, prose and formula. Ideally the AL AST would separate them as different classes of expressions (possibly duplicating some constructs like iter that can occur on both levels). But short of ensuring the layering by construction, it at least needs to be assumed in the renderer logic.

Does that make sense?

Reply:
Aha, thank you for the detailed description, things got much clearer to my mind with the categorization.

I prefer the *or it can legally occur, but then render_expr' needs to render the list as proper prose, e.g., something like "the sequence consisting of ..., ..., and ..."* option. So with the latest commit, I have removed all rendering logic in `render_expr'` that belongs to *none* of the three category, i.e., the ones that produce math-like output in plaintext.

By the fallback logic from `render_expr` to `render_expr'`, when rendering `UnE` in `render_expr'`, it must be that the operator is not a formula. But to make things more clear when *reading* the code, I added a guard indicating that it is the case. And similarly for `BinE` and `IterE`.

```ocaml
(* before *)
| Al.Ast.UnE (op, e) ->
      let sop = render_al_unop op in
      let se = render_expr env e in
      sprintf "%s %s" sop se

(* now *)
| Al.Ast.UnE (op, e) when Option.is_none (al_to_el_unop op) ->
      let sop = render_al_unop op in
      let se = render_expr env e in
      sprintf "%s %s" sop se
```

Plus, I do believe that the invariant you made is true, but I added this final case to `render_expr'` since I do not want the renderer to crash on some unexpected, peculiar expression form.

```ocaml
| _ ->
      let s = Al.Print.string_of_expr expr in
      sprintf "warning: %s was not properly handled by prose backend" s |> prerr_endline;
      s
```

And a final remark, I prefer to keep the option monadic translation of AL-expr-to-EL-exp. First, due to the aforementioned presence of `IterE`. Second, figuring out if an AL expr is of category 1 and actually carrying out the translation requires one to recurse through all the subexpressions. I prefer to do such in a single function `al_to_el_exp`, which returns some EL exp if it belongs to category 1, and none otherwise.